Add Azure OpenAI and OpenRouter Provider Support

bsormagec opened the pull request:
This PR adds support for using Azure OpenAI and OpenRouter as LLM providers within the extension.
  
  - Users can now configure their Azure OpenAI endpoints, deployment names, API versions, and keys in the settings.
  
  - OpenRouter configuration is also added, allowing users to leverage its API.
  - Added OpenRouter and auto-fetch and filter models, including only the free ones!
  
  - The underlying code for managing provider configurations and creating chat models has been updated to handle these new provider types.
  
  fixes #69

alexchenzl commented on pages/side-panel/src/components/ChatHistoryList.tsx:
looks like this change is not necessary

alexchenzl commented on pages/side-panel/src/components/MessageList.tsx:
looks like this change is not necessary

alexchenzl reviewed (commented):
@bsormagec Thank you for your awesome work! Lots of users want the Azure supports!

The following are my feedbacks after some tests:

1. Looks like users can not add more than one Azure OpenAI models in settings. 
2. For OpenRouter provider, I have ever thought of fetching models before, but I decide not to do it in the extension for these considerations:

   a. OpenRouter has too many models, it will make the UI/UX too complex
   
   b. A lot of OpenRouter models are not supported well. For example, openai/gpt-4o-2024-11-20 works well, but openai/gpt-4o not.
   
   c. Users of OpenRouter usually are developers, it's not difficult for them to add models manually.
   d. For a mid-term plan, I will add a backend API for these cloud LLM providers. It will provide a general interface to fetch default models, searching with filters and so on. It will also improve the simplicity of the frontend code in the extension.

3. I hope you could understand my considerations above, would you mind helping to make the following changes:

   a. Set several default models that work well with Nanobrowser in OpenRouter
   
   b. Remove fetching models from OpenRouter, just let users add modes manually if they are not in the default models
   
   c. Add custom headers to identify Nanobrowser in OpenRouter if possible. For your reference: https://github.com/OpenRouterTeam/openrouter-examples/blob/main/examples/langchain/index.ts#L25

bsormagec commented:
Hi @alexchenzl, yeah, I can make those changes too. It takes a little time for me, but yeah, I can. I just needed it yesterday and just made it for myself and then shared it with the community. And I will try to do that your requests and mention you soon. 👍🏼

alexchenzl commented:
> Hi @alexchenzl, yeah, I can make those changes too. It takes a little time for me, but yeah, I can. I just needed it yesterday and just made it for myself and then shared it with the community. And I will try to do that your requests and mention you soon. 👍🏼

Really appreciate your great work and understanding!

alexchenzl reviewed (commented):
1.    Looks like this part was removed:

// O series models have different parameters
  if (isOpenAIOModel(modelConfig.modelName)) {
    args.modelKwargs = {
      max_completion_tokens: maxTokens,
    };
  } else {
  }
  
 2. For Azure OpenAI config, model name is removed. Is there any way to know the model name so that we can apply different parameters like above ?

bsormagec commented:
I am working to add it back on the new revision. 

> 1. Looks like this part was removed:
> 
> // O series models have different parameters if (isOpenAIOModel(modelConfig.modelName)) { args.modelKwargs = { max_completion_tokens: maxTokens, }; } else { }
> 
> 2. For Azure OpenAI config, model name is removed. Is there any way to know the model name so that we can apply different parameters like above ?

bsormagec commented:
Hi @alexchenzl, it looks like this is ready for merge! I’ve added some more features, especially for reasoning models, and updated all the flows as you suggested. Everything seems set for deployment now. I hope the community enjoys it—and I never say no to anyone who wants to thank me! :)

P.S. If you’re feeling extra grateful and want to leave a tip, you can do so on my GitHub Sponsors page. Your tip will not only make my day, but it might also make my coffee cup overflow with joy (and caffeine)! ☕💸

bsormagec commented:
Can you check it @alexchenzl, i am waiting to use it 😅

alexchenzl reviewed (commented):
![image](https://github.com/user-attachments/assets/78280e19-1ed5-4852-8c06-9423dcc61d66)
It's almost done, but I found some minor issues when adding Azure OpenAI provider:

1. There are two * appended to label Endpoint
2. The * appended to label Deployment is not displayed
3. "No deployment names set. Add model names below." Looks like  these words not necessary. The placeholder of input field should be enough
4. What will happen if the deployment name is not a correct model name? For example, "my-gpt-4o" instead of "gpt-4o"

bsormagec commented:
> I also have a question:
> 
> What will happen if the deployment name is not a formal OpenAI model name? Will a deployment name like _my-gpt-4o_ work ?

Hi @alexchenzl 

By default, the model name appears as the deployment name in the Azure Panel. If it changes, the setup will likely not work. If you want to use it here, you should keep it the same; otherwise, you cannot use multiple models under a single endpoint because Azure creates one endpoint for each location. You can access all of them through the main domain by changing only the model information. Therefore, I don't think it's necessary to introduce this complexity into the application right now. If you want to use it here, use it in this way; otherwise, the code will become complex, and the clean code structure will be disrupted, requiring many if blocks. If you say we should switch to that format, then we can, but I think it's beneficial to keep it simple in this way.

alexchenzl commented:
> > I also have a question:
> > What will happen if the deployment name is not a formal OpenAI model name? Will a deployment name like _my-gpt-4o_ work ?
> 
> Hi @alexchenzl
> 
> By default, the model name appears as the deployment name in the Azure Panel. If it changes, the setup will likely not work. If you want to use it here, you should keep it the same; otherwise, you cannot use multiple models under a single endpoint because Azure creates one endpoint for each location. You can access all of them through the main domain by changing only the model information. Therefore, I don't think it's necessary to introduce this complexity into the application right now. If you want to use it here, use it in this way; otherwise, the code will become complex, and the clean code structure will be disrupted, requiring many if blocks. If you say we should switch to that format, then we can, but I think it's beneficial to keep it simple in this way.

I agree with that we should not introduce too much complexity here.  But we can not prevent users from changing default deployment names, then they will encounter issues but don't know the reason.

I suggest to add some notes below the deployment input field or in the input placeholder,  in order to let users know that they should keep the deployment name same as the model name. Then they will know the reason when they encounter issues.

bsormagec commented:
> > > I also have a question:
> > > What will happen if the deployment name is not a formal OpenAI model name? Will a deployment name like _my-gpt-4o_ work ?
> > 
> > 
> > Hi @alexchenzl
> > By default, the model name appears as the deployment name in the Azure Panel. If it changes, the setup will likely not work. If you want to use it here, you should keep it the same; otherwise, you cannot use multiple models under a single endpoint because Azure creates one endpoint for each location. You can access all of them through the main domain by changing only the model information. Therefore, I don't think it's necessary to introduce this complexity into the application right now. If you want to use it here, use it in this way; otherwise, the code will become complex, and the clean code structure will be disrupted, requiring many if blocks. If you say we should switch to that format, then we can, but I think it's beneficial to keep it simple in this way.
> 
> I agree with that we should not introduce too much complexity here. But we can not prevent users from changing default deployment names, then they will encounter issues but don't know the reason.
> 
> I suggest to add some notes below the deployment input field or in the input placeholder, in order to let users know that they should keep the deployment name same as the model name. Then they will know the reason when they encounter issues.

OK, I will try to do it.

alexchenzl reviewed (approved):
Thank you for your awesome work!

bsormagec commented:
I hope everyone loves it! 🤘🏼 @alexchenzl
